docs/source/howto/codes.rst

```diff
@@ -45,8 +45,84 @@ Design guidelines
+    calculation_class = CalculationFactory('arithmetic.add')
+
+We will obviously also need to provide the inputs for the code when running the calculation.
+You can also use ``verdi plugin`` to check this and other relevant information for specific plugins:
+
+.. code-block:: bash
+
+    $ verdi plugin list aiida.calculations arithmetic.add
+
+You will see that 3 inputs nodes are required: two containing the values to add up (``x``, ``y``) and one containing information about the specific code to execute (``code``).
+If you already have these nodes in your database, you can get them by :ref:`querying for them <how-to:data:finding-data>` or using ``orm.load_node(<PK>)``.
+Otherwise, you will need to create them as shown below (note that you `will` need to already have the ``localhost`` computer configured):
+
+.. code-block:: python
+
+    from aiida import orm
+    code_node = orm.Code(remote_computer_exec=[localhost, '/bin/bash'])
+    numx_node = orm.Int(17)
+    numy_node = orm.Int(11)
+
+To provide these as inputs to the calculations, we will now use the ``builder`` object that we can get from the class:
+
+.. code-block:: python
+
+    calculation_builder = calculation_class.get_builder()
+    calculation_builder.code = code_node
+    calculation_builder.x = numx_node
+    calculation_builder.y = numy_node
+
+Now everything is in place and ready to perform the calculation, which can be done in two different ways.
+The first one is blocking and will return a dictionary containing all the output nodes (keyed after their label, so in this case these should be: "remote_folder", "retrieved" and "sum") that you can safely inspect and work with:
+
+.. code-block:: python
+
+    from aiida.engine import run
+    output_dict = run(calculation_builder)
+    sum_node = output_dict['sum']
+
+The second one is non blocking, as you will be submitting it to the daemon and immediately getting control in the interpreter.
+The return value in this case is the actual calculation node that is stored in the database.
+
+.. code-block:: python
+
+    from aiida.engine import submit
+    calculation_node = submit(calculation_builder)
+
+Note that, although you have access to the node, the underlying calculation `process` is not guaranteed to have finished when you get back control in the interpreter.
+In order to keep track of it you can use the verdi command line interface:
+
+.. code-block:: bash
+
+    $ verdi process list
+
+Performing a dry-run
+--------------------
+
+Additionally, you might want to check and verify your inputs before actually running or submitting a calculation.
+You can do so by specifying to use a ``dry_run``, which will create all the input files in a local directory (``submit_test/[date]-0000[x]``) so you can inspect them before actually launching the calculation:
+
+.. code-block:: python
+
+    calculation_builder.metadata.dry_run = True
+    calculation_builder.metadata.store_provenance = False
```

Comment:
Quick question: Does `dry_run = True` not imply `store_provenance = False`?

Reply:
@sphuber I do not know this, honestly I just used the same two lines that were present in the aiida tutorial of 2019.

Reply:
No, dry run means the job does not actually get submitted and starts running, but simply creates the input files in a local folder. It will still create a node in the database and link the inputs up. If you do not want this (which is likely if you are running a dry run) you need to set it

Reply:
Seems like "dry-run" is a bit of a misnomer then, but anyways, doesn't really affect the documentation.

Reply:
why do you think it is a misnomer?

Reply:
Because "dry-run" to me implies that nothing is actually going to happen, i.e., it is perfectly safe to execute without any effects. But I see that in this context it might make sense.